### ISSUE TITLE
[enhancement] option to ignore hard-wrapping and formatting for specified lines

### ISSUE THREAD

**Issue by ajgae** [vim/vim#7714]
# Problem

## Current behaviour

I often write markdown in Vim, intended for use with `pandoc`. I like to have `textwidth=80` in order to get some hard-wrapping. This works nicely most of the time, even with lists (`-`), numbered lists (`1.`, `2.`, ...), and citation blocks (`>`). But when I write [pandoc markdown pipe tables](https://pandoc.org/MANUAL.html#extension-pipe_tables) with long entries, hard-wrapping causes the lines to be broken and this destroys the intended table formatting. For example (with `vim -Nu NONE`, or `vim --clean`, and then `:set textwidth=80`):

```markdown
- this very long list entry, which is over 80 (=`textwidth`) characters long, still looks fine after getting hard-wrapped

| table column 1 | table column 2 | table column 3 |
|---|---|---|
| this is a short entry | this is a medium-sized table entry | this is a long entry in a markdown table which works fine as long as it doesn't get hard-wrapped |
```

after doing `gggqG`, the file contents change to this:

```markdown
- this very long list entry, which is over 80 (=`textwidth`) characters long,
  still looks fine after getting hard-wrapped

| table column 1 | table column 2 | table column 3 | |---|---|---| | this is a
short entry | this is a medium-sized table entry | this is a long entry in a
markdown table which works fine as long as it doesn't get hard-wrapped |
```

At this stage, the table formatting completely wrong, as the adjacent table lines have been considered part of a single paragraph, and were therefore joined before getting hard-wrapped. Even if we ignore the joining-before-getting-hard-wrapped problem (i.e. we don't use `gggqG` but simply just write out the text in insert mode and that's it), the hard-wrapped problem is still there: 

```markdown
- this very long list entry, which is over 80 (=`textwidth`) characters long,
  still looks fine after getting hard-wrapped

| table column 1 | table column 2 | table column 3 |
|---|---|---|
| this is a short entry | this is a medium-sized table entry | this is a long
entry in a markdown table which works fine as long as it doesn't get hard-wrapped |
```

I now have to manually join the table lines together using `J`, and avoid using `gq` on this portion of the file for the rest of the time I spend editing it (which means no quick `gggqG`).

## Expected behaviour

I would like to be able to setup Vim so that certain lines are not hard-wrapped or formatted at all, by specifying that lines starting with (any whitespace and then) `|` should not be hard-wrapped despite `textwidth` being set, and should not be merged when formatting either.

# Proposed solution

I would propose adding something like a `formatignore` string, similar to but distinct from the existing `comments` string, containing a comma-separated list of strings that, if encountered at the beginning of a line, would mean the `textwidth` option should be ignored for that line and it shouldn't be hard-wrapped or formatted. This would fix a similar concern brought up by [this stackexchange post](https://vi.stackexchange.com/q/21272) about the lack of an option to ignore certain lines when formatting, and specifically hard-wrapping.

# Alternatives

The aforementioned [stackexchange thread](https://vi.stackexchange.com/q/21272) contains a suggested [fix](https://vi.stackexchange.com/a/21276), however this seems like just a verbose workaround for an option that should be easily accessible.

I tried adding `|` to the `comments` string and messing with the different options of the `formatioptions` string, and while I was able to fix the "merging adjacent lines" problem for short lines, I couldn't do anything about the hard-wrapping of long lines. This seems like a suboptimal workaround as well.

It may be possible for someone competent in Vim script to fix the issue specifically for markdown tables by modifying the `ftplugin` for markdown, in which case my proposed solution may be overkill. However, not everyone might use this specific markdown pipe table syntax when writing markdown files in Vim, and it would therefore be preferable if the control was in the hands of the user with a dedicated option.

**Comment by brammool** [comment]
Have you tried adding "l" (the letter l) to 'formatoptions'?  It's not perfect, but at least avoids breaking lines that are already long.

**Comment by gdupras** [comment]
> Have you tried adding "l" (the letter l) to 'formatoptions'? It's not perfect, but at least avoids breaking lines that are already long.

This only works in insert mode. Doing something like `gggqG` will still break long lines.

**Comment by brammool** [comment]
True, "l" only helps when inserting text.  For me that is usually sufficient, but it's a primitive method.
Adding a 'formatignore' pattern sounds like a simple method that works in most cases.  e.g. I would set it to "^\s*import" to not format lines starting with "import".  You can use "\|" in a pattern, thus comma's won't be needed.
